### PR TITLE
Slice uid allocator

### PR DESCRIPTION
**Pull request by colin-grapl**
### Which issue does this PR correspond to?
913

### What changes does this PR make to Grapl? Why?
Implements RPCs for creating and incrementing tenant uid namespaces.

### How were these changes tested?
There's integration tests in tests/

### CODE REVIEW

File: src/rust/uid-allocator/src/counters_db.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                     SELECT counter as prev
+                     FROM counters
+                     WHERE counters.tenant_id = $2
+                     LIMIT 1
+                 ) as c
+            WHERE counters.tenant_id = $2
+            RETURNING counter as new, c.prev
+            ",
+            size as i32,
+            tenant_id
```
Comment (wimax-grapl) [inline src/rust/uid-allocator/src/counters_db.rs:42]: Just to make sure, this is multiprocess-safe? I don't really know postgres's lock guarantees

File: src/rust/uid-allocator/src/service.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+impl From<UidAllocatorServiceError> for Status {
+    fn from(err: UidAllocatorServiceError) -> Self {
+        match err {
+            UidAllocatorServiceError::SqlxError(err) => {
+                Status::internal(format!("Internal database error: {}", err))
+            }
+            UidAllocatorServiceError::UnknownTenant(tenant_id) => {
+                Status::not_found(format!("Unknown Tenant: {}", tenant_id))
+            }
+        }
```
Comment (wimax-grapl) [inline src/rust/uid-allocator/src/service.rs:49]: you can use `{err}` btw

File: src/rust/uid-allocator/src/service.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        request: CreateTenantKeyspaceRequest,
+    ) -> Result<CreateTenantKeyspaceResponse, Self::Error> {
+        let tenant_id = request.tenant_id;
+        sqlx::query!(
+            r"INSERT INTO counters (tenant_id, counter) VALUES ($1, 1)
+            ON CONFLICT DO NOTHING;",
+            tenant_id
+        )
+        .execute(&self.allocator.db.pool)
+        .await?;
```
Comment (wimax-grapl) [inline src/rust/uid-allocator/src/service.rs:79]: this feels like it should be on the `db` object

File: src/rust/uid-allocator/tests/uid_allocator.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+async fn test_uid_allocator() -> Result<(), Box<dyn std::error::Error>> {
+    let client_config = uid_allocator::config::UidAllocatorClientConfig::parse();
+
+    let tenant_id = uuid::Uuid::new_v4();
+
+    let endpoint = format!("http://{}", &client_config.uid_allocator_connect_address);
+    let mut allocator_client = CachingUidAllocatorServiceClient::new(
+        UidAllocatorServiceClient::connect(endpoint).await?,
+        100,
+    );
```
Comment (wimax-grapl) [inline src/rust/uid-allocator/tests/uid_allocator.rs:20]: can we maybe provide some prebuilt `fn (UidAllocatorClientConfig) -> CachingUidAllocatorServiceClient` ?

in fact, the `allocation_size` in that ClientConfig is never used, strongly suggesting this function should exist

**Review by wimax-grapl — COMMENTED** [review]
mostly looks good

File: src/rust/uid-allocator/src/allocator.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+};
+
+/// PreAllocation is intended to be a struct large enough to represent many allocations worth
+/// of Uid ranges. When clients request a new allocation, first it is taken from the PreAllocation,
+/// so that we only reach out to the database when the PreAllocation is exhausted.
+#[derive(Clone, Copy)]
+pub struct PreAllocation {
+    pub start: u64,
+    pub current: u64,
+    pub end: u64,
```
[the remark this answers is not in the document]
Comment (colin-grapl) [reply src/rust/uid-allocator/src/allocator.rs:22]: I believe it is inclusive. I will verify and add a comment.

File: src/rust/uid-allocator/src/counters_db.rs (hunk 1 above)
Comment (colin-grapl) [reply src/rust/uid-allocator/src/counters_db.rs:42]: I should add FOR UPDATE to the select just in case.

**Review by wimax-grapl — APPROVED** [review]
Approving conditionally, in that colin will fix the "add a (config) -> client fn" later this week

**Pull request merged**